clients/include/testing_reduction_batched.hpp

```diff
@@ -0,0 +1,201 @@
+/* ************************************************************************
+ * Copyright 2018-2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+
+#include "cblas_interface.hpp"
+#include "norm.hpp"
+#include "rocblas.hpp"
+#include "rocblas_cblas.hpp"
+#include "rocblas_init.hpp"
+#include "rocblas_math.hpp"
+#include "rocblas_random.hpp"
+#include "rocblas_test.hpp"
+#include "rocblas_vector.hpp"
+#include "unit.hpp"
+#include "utility.hpp"
+
+template <typename T, typename R>
+using rocblas_reduction_batched_t = rocblas_status (*)(rocblas_handle  handle,
+                                                       rocblas_int     n,
+                                                       const T* const* x,
+                                                       rocblas_int     incx,
+                                                       rocblas_int     batch_count,
+                                                       R*              result);
+
+template <typename T, typename R>
+void template_testing_reduction_batched_bad_arg(const Arguments&                  arg,
+                                                rocblas_reduction_batched_t<T, R> func)
+{
+    rocblas_int N = 100, incx = 1, batch_count = 5;
+
+    rocblas_local_handle handle;
+
+    //
+    // allocate memory on device
+    //
+    device_batch_vector<T> dx(N, incx, batch_count);
+    CHECK_HIP_ERROR(dx.memcheck());
+
+    R h_rocblas_result;
+
+    EXPECT_ROCBLAS_STATUS(func(handle, N, nullptr, incx, batch_count, &h_rocblas_result),
+                          rocblas_status_invalid_pointer);
+
+    EXPECT_ROCBLAS_STATUS(func(handle, N, dx, incx, batch_count, nullptr),
+                          rocblas_status_invalid_pointer);
+
+    EXPECT_ROCBLAS_STATUS(func(nullptr, N, dx, incx, batch_count, &h_rocblas_result),
+                          rocblas_status_invalid_handle);
+}
+
+template <typename T, typename R>
+void template_testing_reduction_batched(const Arguments&                  arg,
+                                        rocblas_reduction_batched_t<T, R> func,
+                                        void (*CBLAS_FUNC)(rocblas_int, const T*, rocblas_int, R*))
+{
+    rocblas_int          N = arg.N, incx = arg.incx, batch_count = arg.batch_count;
+    rocblas_stride       stride_x = arg.stride_x;
+    double               rocblas_error_1, rocblas_error_2;
+    rocblas_local_handle handle;
+
+    if(batch_count < 0)
+    {
```

Comment:
**This is being refactored as part of the argument error check changes.** I have already made changes to {IAMAX,IAMIN,ASUM,NRM2} batched tests; similar changes will be needed here now or later.

Basically:

if `N<=0 || incx <= 0 || batch_count == 0`, then the result is the `default_value` of the reduction, i.e., the reduction when applied to a vector of 0 elements. There is no error condition (return `rocblas_status_success`).

If `N > 0 && incx > 0 && batch_count < 0`, it is a `rocblas_status_invalid_size`.

So the normal flow is like:
```
if (N <= 0 || incx <=0 || batch_count <= 0)
{
   // Allocate safe size
   // ..
   EXPECT_ROCBLAS_STATUS(func(handle, N, dx.ptr_on_device(), incx, batch_count, dr).
                              N > 0 && incx > 0 && batch_count < 0 ? rocblas_status_invalid_size :
                              rocblas_status_success);
}
```

Reply:
Fixed.